src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_AssignmentOperator.cs

```diff
@@ -229,6 +229,11 @@ private BoundExpression VisitAssignmentOperator(BoundAssignmentOperator node, bo
 
                 case BoundKind.DiscardExpression:
                     {
+                        if (isRef && rewrittenRight is BoundArrayAccess arrayAccess)
+                        {
+                            return arrayAccess.Update(isRef: true);
```

Comment:
>return arrayAccess.Update(isRef: true); [](http://example.com/codeflow?start=28&length=39)

Is there a good reason to adjust behavior for this scenario? Do we have any debug/release difference in behavior without the change?

Reply:
> Do we have any debug/release difference in behavior without the change?

No, there is no debug/release difference neither before nor after the change. Before the change both debug and release emitted `ldelem` or similar; after the change both debug and release emit `ldelema`.

> Is there a good reason to adjust behavior for this scenario?

Using `ref array[x]` and then not using the result (either by not using the local or discarding it explicitly) feels like something that should behave equivalently.

Although now I realized that in case of a ref-readonly assignment we also emit `readonly.ldelema` which doesn't do the runtime element type check, and discards could be considered readonly as well.

So I agree, I will revert this part of the change.